api_volontaria/apps/position/models.py

```diff
@@ -0,0 +1,184 @@
+            return True
+        else:
+            return False
+
+    @staticmethod
+    def has_update_permission(request):
+        if request.user.is_staff:
+            return True
+        else:
+            return False
+
+    @authenticated_users
+    def has_object_destroy_permission(self, request):
+        if request.user.is_staff:
+            return True
+        else:
+            return False
+
+    @authenticated_users
+    def has_object_update_permission(self, request):
+        if request.user.is_staff:
+            return True
+        else:
+            return False
+
+
+class Application(models.Model):
+    """
+    This class represents an application made by volunteer for a given position.
+    """
+
+    APPLICATION_UNDER_EXAMINATION = 'UNDER_EXAMINATION'
+    APPLICATION_ACCEPTED = 'ACCEPTED'
+    APPLICATION_DECLINED = 'DECLINED'
+
+    APPLICATION_CHOICES = (
+        (APPLICATION_UNDER_EXAMINATION, _('Under examination')),
+        (APPLICATION_ACCEPTED, _('Accepted')),
+        (APPLICATION_DECLINED, _('Declined')),
+    )
+
+    class Meta:
+        verbose_name = _("Application")
+        verbose_name_plural = _("Applications")
+
+    position = models.ForeignKey(
+        Position,
+        verbose_name=_("Position"),
+        blank=False,
+        on_delete=models.CASCADE,
+    )
+    
+    applied_on = models.DateTimeField(
+        verbose_name=_("Applied on"),
+        auto_now_add=True,
+    )
+    
+    motivation = models.TextField(
+        verbose_name="Motivation",
+    )
+
+    application_status = models.CharField(
+        verbose_name=_("Application status"),
+        max_length=100,
+        choices=APPLICATION_CHOICES,
+        default=APPLICATION_UNDER_EXAMINATION
+    )
+
+    user = models.ForeignKey(
+        User,
+        related_name='applications',
+        on_delete=models.CASCADE,
+    )
+
+    def __str__(self):
+        return self.application.name
+        # return self.position.name
+
+    @staticmethod
+    def has_destroy_permission(request):
+        if request.user.is_staff:
+            return True
+        else:
+            return False
+
+    @staticmethod
+    def has_update_permission(request):
+        if request.user.is_staff:
+            return True
+        else:
+            return False
+
+    @staticmethod
+    def has_list_permission(request):
+        return True
+
+    @staticmethod
+    def has_create_permission(request):
+        return True
+
+    @authenticated_users
+    def has_object_update_permission(self, request):
+        if self.user == request.user:
+            return True
+        if request.user.is_staff:
+            return True
+        else:
+            return False
+
+    @authenticated_users
+    def has_object_destroy_permission(self, request):
+        if self.user == request.user:
+            return True
+        if request.user.is_staff:
+            return True
+        else:
+            return False
```

Comment:
I think that user should be able to destroy their own participation IF the participation is under examination only. What do you think ? It could help admin keep track of the process and not give the sentiment that application desapear after review.

Reply:
@RignonNoel 

Good idea.

So, I updated the code accordingly. But, when writing a test "test_delete_accepted_application", I've just realized that the current solution may actually prevent a user from deleting its own application, regardless of the application status. 
Which is not what we want: in models.py, the Application class is supposed to give user the right to delete its own application.
Do we have a bug here?

In test_view_applications.py:

`def test_delete_application(self):
        """
        Ensure we can't delete an application if we are a simple user.
        """
        self.client.force_authenticate(user=self.user)

        response = self.client.patch(
            reverse(
                'application-detail',
                kwargs={
                    'pk': self.application.id
                },
            )
        )

        content = json.loads(response.content)

        self.assertEqual(response.status_code, status.HTTP_403_FORBIDDEN)
        self.assertEqual(
            content,
            {
                'detail': 'You do not have permission to perform this action.'
            }
        )`



But, in models.py:

 @authenticated_users
    def has_object_update_permission(self, request):
        if self.user == request.user:
            return True
        if request.user.is_staff:
            return True
        else:
            return False